src/main/kotlin/tech/relaycorp/relaynet/Certificate.kt

```diff
@@ -0,0 +1,106 @@
+
+class Certificate constructor (certificateHolder: X509CertificateHolder?) {
+
+    companion object {
+        val MAX_PATH_LENGTH_CONSTRAINT = 2
+        val DEFAULT_ALGORITHM = "SHA256WithRSAEncryption"
+
+        @Throws(CertificateError::class)
+        fun issue(options: FullCertificateIssuanceOptions): Certificate {
+            val start = options.validityStartDate
+            val end = options.validityEndDate
+            // validate inputs
+            if (start >= end) {
+                throw CertificateError("The end date must be later than the start date")
+            }
+
+            val issuer = X500Name.getInstance(options.commonName)
+            val serial = options.serialNumber
+
+            val subject = options.commonName
+            val pubkey = options.subjectPublicKey
+            val subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(pubkey.getEncoded())
+
+            val signatureAlgorithm = DefaultSignatureAlgorithmIdentifierFinder().find(DEFAULT_ALGORITHM)
+            val digestAlgorithm = DefaultDigestAlgorithmIdentifierFinder().find(signatureAlgorithm)
+            val privateKeyParam: AsymmetricKeyParameter = PrivateKeyFactory.createKey(options.issuerPrivateKey.encoded)
+            val contentSignerBuilder = BcRSAContentSignerBuilder(signatureAlgorithm, digestAlgorithm)
+            val signerBuilder = contentSignerBuilder.build(privateKeyParam)
+
+            val builder = X509v3CertificateBuilder(issuer,
+                    serial.toBigInteger(), Date.valueOf(start.toLocalDate()), Date.valueOf(end.toLocalDate()), Locale.ENGLISH, subject, subjectPublicKeyInfo)
+
+            return Certificate(builder.build(signerBuilder))
+        }
+
+        @Throws(CertificateError::class)
+        fun buildX500Name(cName: String): X500Name {
+            if (cName.length <= 0) {
+                throw CertificateError("Invalid CName in X500 Name")
+            }
+            val builder = X500NameBuilder(BCStyle.INSTANCE)
+            builder.addRDN(BCStyle.C, cName)
+            return builder.build() ?: throw CertificateError("Invalid X500 Name")
+        }
+    }
+}
+
+data class FullCertificateIssuanceOptions(
+    var commonName: X500Name?,
+    var issuerPrivateKey: PrivateKey,
+    var subjectPublicKey: PublicKey,
+    var serialNumber: Long = generateRandom64BitValue(),
+    var validityStartDate: LocalDateTime = LocalDateTime.now(),
+    var validityEndDate: LocalDateTime = validityStartDate.plusMonths(1),
+    var isCA: Boolean? = false,
+    var issuerCertificate: Certificate?,
+    var pathLenConstraint: Int = MAX_PATH_LENGTH_CONSTRAINT
+)
+
+class Keys {
+
+    companion object {
+        private val DEFAULT_KEY_SIZE: Int = 2048
+        fun generateRSAKeyPair(modulus: Int): KeyPair {
+            val keyGen = KeyPairGenerator.getInstance("RSA")
+            keyGen.initialize(maxOf(modulus, DEFAULT_KEY_SIZE)) // `modulus` should be >= 2048 and default to 2048
```

Comment:
If the caller attempts to use an insecure modulus (anything under 2048), I'd suggest to throw an exception instead of gracefully using 2048.

Apart from being explicit, it may also save some troubleshooting down the line if they store the key in some data store where it'd get truncated to the expected length (not that I think that'd apply to Android, but it might apply to server-side apps using this library).